datasetinsights/data/download.py

```diff
@@ -0,0 +1,66 @@
+import logging
+from pathlib import Path
+
+import requests
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+
+from .exceptions import DownloadError
+
+logger = logging.getLogger(__name__)
+
+# Timeout of requests (in seconds)
+DEFAULT_TIMEOUT = 1800
+# Retry after failed request
+DEFAULT_MAX_RETRIES = 5
+
+
+class TimeoutHTTPAdapter(HTTPAdapter):
+    def __init__(self, timeout, *args, **kwargs):
+        self.timeout = timeout
+        super().__init__(*args, **kwargs)
+
+    def send(self, request, **kwargs):
+        kwargs["timeout"] = self.timeout
+        return super().send(request, **kwargs)
+
+
+def download_file(source_uri: str, dest_path: str, use_cache: bool = True):
+    """Download a file specified from a source uri
+
+    Args:
+        source_uri (str): source url where the file should be downloaded
+        dest_path (str): destination path of the file
+        use_cache (bool): use_cache (bool): use cache instead of
+                re-download if file exists
+
+    Returns:
+        String of destination path.
+    """
+    dest_path = Path(dest_path)
+    if dest_path.exists() and use_cache:
+        return dest_path
+
+    logger.debug(f"Trying to download file from {source_uri} -> {dest_path}")
+    adapter = TimeoutHTTPAdapter(
+        timeout=DEFAULT_TIMEOUT, max_retries=Retry(total=DEFAULT_MAX_RETRIES)
+    )
+    with requests.Session() as http:
+        http.mount("https://", adapter)
+        try:
+            response = http.get(source_uri)
+            response.raise_for_status()
+        except requests.exceptions.RequestException as ex:
+            logger.error(ex)
+            err_msg = (
+                f"The request download from {source_uri} -> {dest_path} can't "
+                f"be completed."
+            )
+
+            raise DownloadError(err_msg)
+        else:
+            dest_path.parent.mkdir(parents=True, exist_ok=True)
+            with open(dest_path, "wb") as f:
+                f.write(response.content)
+
+    return dest_path
```

Comment:
Can we rename this file name, like `http_download.py`? It's wired to have the same name with `datasetinsights/data/simulation/download.py`.

Reply:
yes, we can do that. But this is contains generic download function on the other hand, `datasetinsights/data/simulation/download.py` contains download functions related to USIM so we can also change that to `usim_download.py`

Reply:
Ok. Sounds good!

Reply:
I talked with YC about this, and we think it's fine to have `download.py` in both places.

Reply:
Okay. I'm okay with that.